src/helpers/Monitor.cpp

```diff
@@ -21,4 +21,11 @@ void SMonitor::registerListeners() {
 
         description = desc;
     });
+
+    output->setGeometry([this](CCWlOutput* r, int32_t x, int32_t y, int32_t width_mm, int32_t height_mm, int32_t subpixel, const char* make, const char* model,
+                               int32_t transform_) { //
+        transform = (wl_output_transform)transform_;
+        if (transform == WL_OUTPUT_TRANSFORM_90 || transform == WL_OUTPUT_TRANSFORM_270)
+            std::swap(size.x, size.y);
```

Comment:
works perfectly fine on my end, but i will implement what i understand of what you just said.

Reply:
Yeah no I'm sorry but i am really struggling to understand what the issue is at all.

The wallpaper will react correctly to literally any geometry or mode changes i throw at my desktop monitor through hyprctl, and I just spent the last like 10 minutes setting different resolutions, transforms, or combinations of both and in every case the wallpaper was correct. (to clarify, i changed resolution with no transform change,  resolution changed with transform change, and transform change but no resolution change, every single time it was correct.)

```
❯ ./hyprpaper
[LOG] Welcome to hyprpaper!
built from commit ecab9e72910dfacd5d668e8722a87aab47471e29 (    implement wl_output_transform)
[LOG] Cleaned old hyprpaper preloads (2), removing 23.0MB
[LOG] Preloaded target /home/isaac/Documents/houseonthesideofalake.jpg in 347.79ms -> Pixel size: [5967, 3978]
[LOG] Buffer created for target /home/isaac/Documents/houseonthesideofalake.jpg, Shared Memory usage: 8.3MB
[LOG] Buffer created for target /home/isaac/Documents/houseonthesideofalake.jpg, Shared Memory usage: 14.7MB
[LOG] Total SM usage for all buffers: 23.0MB
[LOG] hyprpaper socket started at /run/user/1000/hypr/e2c78c00e50a0835c3830c8a0cd82049dc5700b8_1734460376_53452080/.hyprpaper.sock (fd: 5)
[LOG] configure for eDP-1
[LOG] handlePreferredScale: 1.00 for 3fddd4f0
[LOG] configure for eDP-1
[LOG] configure for HDMI-A-1
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for eDP-1: /home/isaac/Documents/houseonthesideofalake.jpg at [0.00, -310.78], scale: 0.32 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 1920x1080 for 3fddd4f0
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [0.00, -310.78], scale: 0.43 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 2560x1440 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fddd4f0
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [0.00, -310.78], scale: 0.32 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 1920x1080 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Buffer created for target /home/isaac/Documents/houseonthesideofalake.jpg, Shared Memory usage: 8.3MB
[LOG] Total SM usage for all buffers: 31.3MB
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [-1864.69, 0.00], scale: 0.48 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 1080x1920 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [0.00, -310.78], scale: 0.43 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 2560x1440 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Buffer created for target /home/isaac/Documents/houseonthesideofalake.jpg, Shared Memory usage: 14.7MB
[LOG] Total SM usage for all buffers: 46.1MB
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [-1864.69, 0.00], scale: 0.64 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 1440x2560 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [0.00, -310.78], scale: 0.43 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 2560x1440 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [-1864.69, 0.00], scale: 0.64 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 1440x2560 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [0.00, -310.78], scale: 0.43 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 2560x1440 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [-1864.69, 0.00], scale: 0.64 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 1440x2560 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Buffer created for target /home/isaac/Documents/houseonthesideofalake.jpg, Shared Memory usage: 8.3MB
[LOG] Total SM usage for all buffers: 54.4MB
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [-994.50, 0.00], scale: 0.36 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 1440x1440 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [0.00, -310.78], scale: 0.43 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 2560x1440 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [-1864.69, 0.00], scale: 0.64 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 1440x2560 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
[LOG] configure for HDMI-A-1
[LOG] Image data for HDMI-A-1: /home/isaac/Documents/houseonthesideofalake.jpg at [0.00, -310.78], scale: 0.43 (original image size: [5967, 3978])
[LOG] Submitting viewport dest size 2560x1440 for 3fdde300
[LOG] handlePreferredScale: 1.00 for 3fdde300
```

Reply:
If you can provide me with a clearer example example or, even better, how to repro the issue you're talking about then ill happily do it but otherwise what is the point, this works perfectly fine from what i can see.

Reply:
Next day after sleep:

If you are referring to some sort of edge case where by, upon initialisation of the hyprpaper client, the initial mode is advertised by the compositor (potentially not hyprland) AFTER the geometry (this is possible (?) as wayland doc clarifies that these events are not mutually exclusive), then what you are saying makes perfect sense.

However, from what i have seen && although not guaranteed, the mode seems to be advertised every time before the geometry in hyprland at least. POTENTIALLY other compositor implementations are set up in a way which, or are more prone to, triggering this sort of edge case. If so, what you are saying makes perfect sense.

Furthermore, there are a lot of factors that may also come into play here such as the depths of how the wayland protocol works (i have no idea what a damage request is for example) and, conventionally, how modeseting is handled by wayland compositors.

Really though, before i drop everything and rectify this, I would appreciate a clearer explanation/example of what you mean and the consequences of handling geometry events and transforms like this.

Reply:
okay, got it :)